route: Introduce `state: ignore` for route

cathay4t opened the pull request:
Given a system with ignored interface holding static routes,
When applying desire state retrieved by `nmstatectl show` or
`libnmstate.show()`,
The nmstate will fail with error:

```
    Desired route destination: 2001:db8:e::/64 next-hop-interface: eth1
    next-hop-address: 2001:db8:1::2 table-id: 254 not found after applyVg
```

This is caused by nmstate removed ignored interface during merging
desired state with current.

To fix this issue, this patch introduced `state: ignore` to route entry,
for `libnmstate.show()` and `nmstatectl show`, routes next hop to
ignored interface is now marked with `state: ignore` which will be
ignored during apply action.

Integration test cases included.

Resolves: https://issues.redhat.com/browse/RHEL-107130

rajeshrah22 commented on rust/src/lib/route.rs:
```diff
@@ -621,6 +641,21 @@ impl MergedRoutes {
+                if !iface_type.is_userspace() {
+                    Some(name.as_str())
+                } else {
+                    None
```
Why only include userspace interfaces in `ignored_ifaces`? Isn't `Ethernet` type interface a kernel interface?

cathay4t replied:
User space interface cannot hold routes.

cathay4t replied:
Kernel interface means have kernel representative  (interface name or at least an interface index number), containing ethernet, wifi, vlan, bond, linux bridge and etc.

User space interface means they only exist in user space tool's config or database, for example OVS bridge itself only exist in OVSDB.

ihuguet commented:
I don't think that we should ignore the routes. If the user has defined a route in the desired state and we cannot configure it, then we must raise an error. What we can do is to improve the error message to explain better the reason, in any case, but not ignoring the route.

cathay4t commented:
> I don't think that we should ignore the routes. If the user has defined a route in the desired state and we cannot configure it, then we must raise an error. What we can do is to improve the error message to explain better the reason, in any case, but not ignoring the route.

Nmstate promised to support `sudo nmstatectl show| sudo nmstatectl apply -`. We do not state for a route indicate `ignore` or not, I think it is correct to ignoring routes next hop to ignored interface.

ihuguet commented:
> Nmstate promised to support `sudo nmstatectl show| sudo nmstatectl apply -`. We do not state for a route indicate `ignore` or not, I think it is correct to ignoring routes next hop to ignored interface.

Nmstate also promised to validate that the desired state is configured. With this change, in order to support `show | apply` we make worse job at applying a desired state written by the user. If the user writes a yaml, and the yaml cannot be applied, then the expectation is to get an error.

In any case, what we can do is *not* to apply them, but validate them. That way, if we are doing `show | apply` it should work because the routes were already present. If we are doing `apply <user_written_yaml>` it will fail the validation. It will cause that a yaml saved from a previous `nmstatectl show` may fail to apply later, but IMO that's right.

If that's not acceptable, we should just fail directly. Currently, if we try to configure an interface that is NM-unmanaged, we don't just ignore it, and it fails. The same we should do for routes.

Or maybe we should add a `state: ignore` value to routes, and print it when doing `nmstatectl show`.

cathay4t commented:
Let me introduce `state: ignore` for routes

cathay4t commented:
New patch uploaded. Please review.

ihuguet commented on tests/integration/nm/route_test.py:
```diff
@@ -384,3 +388,38 @@ def test_route_delete_next_hop_iface_by_profile_name_no_leftover(
+    ):
+        for route in routes:
+            if route[Route.NEXT_HOP_INTERFACE] == DUMMY1:
+                assert route[Route.STATE] == Route.STATE_IGNORE
```
We should check that the route is printed. This is only checking that it has `state: ignore` IF it's printed. Nothing happens if it's not printed.

cathay4t replied:
Test case amended.

ihuguet commented on tests/integration/route_test.py:
```diff
@@ -2437,3 +2437,71 @@ def test_add_route_with_advmss(eth1_up):
+
+# https://issues.redhat.com/browse/RHEL-107130
+@pytest.mark.tier1
+def test_apply_ignored_routes_to_ignored_iface(eth1_static_ip):
```
What happens if you try to apply a route without state to an ignored interface? We should add a test to see that an error is raised.

And what happens if you apply a route with `state: absent`? A test is also desirable (if it doesn't exist yet)

cathay4t replied:
Both test cases included.

ihuguet commented on rust/src/lib/query_apply/route.rs:
```diff
@@ -213,3 +213,35 @@ impl RouteEntry {
+            })
+            .collect();
+
+        for rts in [self.running.as_mut(), self.config.as_mut()]
```
`routes.running` is ignored in apply, is there any point in marking them with `state: ignore`?

cathay4t replied:
In case in the future we will use running config in follow up processing. If we don't do it here, it will be very hard to debug.

ihuguet commented on rust/src/lib/route.rs:
```diff
@@ -106,6 +106,15 @@ impl Routes {
     }
 
+    pub(crate) fn remove_ignored_routes(&mut self) {
+        for rts in [self.running.as_mut(), self.config.as_mut()]
```
Same

cathay4t committed:
route: Introduce `state: ignore` for route

Given a system with ignored interface holding static routes,
When applying desire state retrieved by `nmstatectl show` or
`libnmstate.show()`,
The nmstate will fail with error:

```
    Desired route destination: 2001:db8:e::/64 next-hop-interface: eth1
    next-hop-address: 2001:db8:1::2 table-id: 254 not found after applyVg
```

This is caused by nmstate removed ignored interface during merging
desired state with current.

To fix this issue, this patch introduced `state: ignore` to route entry,
for `libnmstate.show()` and `nmstatectl show`, routes next hop to
ignored interface is now marked with `state: ignore` which will be
ignored during apply action.

Integration test cases included.

Resolves: https://issues.redhat.com/browse/RHEL-107130

Signed-off-by: Gris Ge <fge@redhat.com>
